channels/client/addin.c

```diff
@@ -390,3 +390,186 @@ PVIRTUALCHANNELENTRY freerdp_channels_load_static_addin_entry(LPCSTR pszName, LP
+channel_client_thread_proc(LPVOID userdata)
+{
+	UINT error = CHANNEL_RC_OK;
+	wStream* data;
+	wMessage message;
+	channel_client_thread_proc_args *args = userdata;
+	while (1)
+	{
+		if (!MessageQueue_Wait(args->queue))
+		{
+			WLog_ERR(TAG, "MessageQueue_Wait failed!");
+			error = ERROR_INTERNAL_ERROR;
+			break;
+		}
+		if (!MessageQueue_Peek(args->queue, &message, TRUE))
+		{
+			WLog_ERR(TAG, "MessageQueue_Peek failed!");
+			error = ERROR_INTERNAL_ERROR;
+			break;
+		}
+
+		if (message.id == WMQ_QUIT)
+			break;
+
+		if (message.id == 0)
+		{
+			data = (wStream*) message.wParam;
+
+			if ((error = args->msg_handler(args->userdata, data)))
+			{
+				WLog_ERR(TAG, "msg_handler failed with error %"PRIu32"!", error);
+				break;
+			}
+		}
+	}
+	ExitThread(error);
+	if (args)
+	{
+		free (args);
+	}
+	return error;
+}
+
+/*  Create message queue and thread or not, depending on settings */
+void
+channel_client_create_handler(rdpContext *ctx, HANDLE thread, LPVOID userdata,
+	UINT (*msg_handler)(LPVOID userdata,  wStream* data), wMessageQueue* queue)
+{
+	channel_client_thread_proc_args *args =
+		calloc (1, sizeof(channel_client_thread_proc_args));
+	args->msg_handler = msg_handler;
+	args->queue = queue;
+	args->userdata = userdata;
+	if (!(ctx->settings->ThreadingFlags
+		& THREADING_FLAGS_DISABLE_THREADS))
+	{
+		if (!(thread = CreateThread(NULL, 0, channel_client_thread_proc,
+	                                     (void*) args,
+	                                     0, NULL)))
+		{
+			WLog_ERR(TAG, "CreateThread failed!");
+			MessageQueue_Free(queue);
+			queue = NULL;
+		}
+	}
+	else
+	{
+		free (args);
+	}
+}
+/* post a message in the queue or directly call the processing handler */
+UINT
+channel_client_post_message(rdpContext *ctx, LPVOID pData, UINT32 dataLength,
+	UINT32 totalLength, UINT32 dataFlags, wStream* channel_data_in,
+	wMessageQueue* queue, UINT (*msg_handler)(LPVOID userdata,  wStream* data),
+	LPVOID userdata)
+{
+	wStream* data_in;
+
+	if ((dataFlags & CHANNEL_FLAG_SUSPEND) || (dataFlags & CHANNEL_FLAG_RESUME))
+	{
+		return CHANNEL_RC_OK;
+	}
+
+	if (dataFlags & CHANNEL_FLAG_FIRST)
+	{
+		if (channel_data_in)
+			Stream_Free(channel_data_in, TRUE);
+
+		channel_data_in = Stream_New(NULL, totalLength);
+	}
+
+	if (!(data_in = channel_data_in))
+	{
+		WLog_ERR(TAG, "Stream_New failed!");
+		return CHANNEL_RC_NO_MEMORY;
+	}
+
+	if (!Stream_EnsureRemainingCapacity(data_in, dataLength))
+	{
+		Stream_Free(channel_data_in, TRUE);
+		channel_data_in = NULL;
+		return CHANNEL_RC_NO_MEMORY;
+	}
+
+	Stream_Write(data_in, pData, dataLength);
+
+	if (dataFlags & CHANNEL_FLAG_LAST)
+	{
+		if (Stream_Capacity(data_in) != Stream_GetPosition(data_in))
+		{
+			WLog_ERR(TAG, "cliprdr_plugin_process_received: read error");
+			return ERROR_INTERNAL_ERROR;
+		}
+
+		channel_data_in = NULL;
+		Stream_SealLength(data_in);
+		Stream_SetPosition(data_in, 0);
+
+		if (!MessageQueue_Post(queue, NULL, 0, (void*) data_in, NULL))
```

Comment:
Just call the handler directly if no threads are used? No queue necessary in that case...